File: src/modules/page_parameters.py
from PySide6.QtWidgets import (QWidget, QVBoxLayout, QScrollArea, QFormLayout, 
                              QPushButton, QLineEdit, QMessageBox, QInputDialog, 
                              QTabWidget, QApplication, QLabel)
from PySide6.QtCore import QFile, QTextStream, QObject, Signal, QIODevice
import sys
from .Ui_PageParameters import Ui_PageParameters  # 假设通过 pyside6-uic 生成的 UI 文件
from modules import Settings
from .utils import PressableLabel, IniEntry, FormRow
from .signals import signals
from .app_config import config_manager
from .base_page import BasePage

class PageParameters(BasePage):
    ui: Ui_PageParameters
    UI_CLASS = Ui_PageParameters

    def _init_state(self) -> None:
        self.filename = str(config_manager.path)
        self.groupLayouts = {}
        self.formRows = {}
        self.groupOrder = []
        self.orderedEntries = []
        self.selectedGroup = ""
        self.oldSelectedGroup = ""
        self.selectedRowIndex = -1

    def _setup_ui(self) -> None:
        pass

    def _bind_signals(self) -> None:
        signals.app_close.connect(self.on_close)
                
    def _bind_events(self) -> None:
        self.ui.btn_load_parm.clicked.connect(self.reload_settings)
        self.ui.btn_save_parm.clicked.connect(self.save_settings)
        self.ui.tabWidget.currentChanged.connect(lambda index: self.on_tab_changed(index))
        self.ui.btn_add.clicked.connect(lambda: self.add_key_value_row())
        self.ui.btn_insert.clicked.connect(lambda: self.insert_key_value_row())
        self.ui.btn_delete.clicked.connect(lambda: self.delete_key_value_row())
        self.ui.btn_up.clicked.connect(lambda: self.move_key_value_row(-1))
        self.ui.btn_down.clicked.connect(lambda: self.move_key_value_row(1))

    def _on_ready(self) -> None:
        self.load_settings()
        config_manager.batch_changed.emit()

    def reload_settings(self) -> None:
        try:
            config_manager.load(self.filename, emit_changes=True)
        except (OSError, TypeError, ValueError) as exc:
            QMessageBox.warning(self, "错误", f"无法加载配置文件: {exc}")
            return
        self.load_settings()

    def load_settings(self):
        self.parse_init_file(self.filename)
        self.ui.tabWidget.clear()
        self.groupLayouts.clear()
        self.formRows.clear()
        self.groupOrder.clear()

        for entry in self.orderedEntries:
            if entry.group not in self.groupLayouts:
                page = QWidget()
                outerLayout = QVBoxLayout(page)
                scroll = QScrollArea()
                scroll.setWidgetResizable(True)
                scroll.setStyleSheet("QScrollArea { border: none; }")
                content = QWidget()
                content.setObjectName("formContainer")  # 设置对象名称，方便样式表选择
                form = QFormLayout(content)
                # 设置样式表
                content.setStyleSheet(Settings.PARAMTER_FORM_STYLE)
                scroll.setWidget(content)
                outerLayout.addWidget(scroll)
                self.ui.tabWidget.addTab(page, entry.group)
                self.ui.tabWidget.currentChanged.connect(lambda index, g=entry.group: 
                    setattr(self, "selectedGroup", self.ui.tabWidget.tabText(index)))
                self.groupLayouts[entry.group] = form
                self.groupOrder.append(entry.group)

            # label = PressableLabel(entry.key)
            # label.setObjectName(entry.key)
            # edit = QLineEdit(entry.value)
            # self.bind_label_events(label)
            # self.bind_edit_events(edit, entry)
            self.formRows.setdefault(entry.group, []).append(self.createFormRow(entry.group, entry.key, entry.value))
            self.refresh_group_layout(entry.group)

    def on_tab_changed(self, index):
        self.oldSelectedGroup = self.selectedGroup
        self.selectedGroup = self.ui.tabWidget.tabText(index)
        self.selectedRowIndex = -1
        self.clear_old_group_style()
        for i in range(self.ui.tabWidget.count()):
            self.ui.tabWidget.tabBar().setStyleSheet("")
        self.ui.tabWidget.tabBar().setStyleSheet(Settings.WIDGET_TAB_STYLE)  # 示例样式

    def bind_label_events(self, label:PressableLabel):
        label.clicked.connect(lambda: self._onLabelClicked(label))
        label.doubleClicked.connect(lambda: self._on_label_double_clicked(label))

    def bind_edit_events(self, edit:QLineEdit, entry:IniEntry):
        edit.returnPressed.connect(lambda: self._apply_entry(IniEntry(entry.group, entry.key, edit.text())))

    def _apply_entry(self, entry: IniEntry) -> bool:
        try:
            return config_manager.set_entry(entry)
        except (TypeError, ValueError) as exc:
            QMessageBox.warning(self, "参数错误", f"{entry.group}.{entry.key}: {exc}")
            return False

    def _onLabelClicked(self, label):
        self.selectedRowIndex = self.find_row_index(self.selectedGroup, label.objectName())
        self.hightlight_selected_row()

    def _on_label_double_clicked(self, label):
        self.selectedRowIndex = self.find_row_index(self.selectedGroup, label.objectName())
        self.hightlight_selected_row()
        newText, ok = QInputDialog.getText(self, "修改标签", "<span style='color: green; font-weight: bold;'>请输入新的标签名:</span>", QLineEdit.Normal, label.text())
        layout = self.groupLayouts.get(self.selectedGroup)
        if not layout:
            return
        # 遍历布局中的子项检查是否已存在同名标签
        has_duplicate = False
        for i in range(layout.rowCount()):
            label_item = layout.itemAt(i, QFormLayout.LabelRole)
            if label_item and label_item.widget():
                if label_item.widget().objectName() == newText and label_item.widget() != label:
                    has_duplicate = True
                    break
        if has_duplicate:
            QMessageBox.warning(self, "错误", "<span style='color: green; font-weight: bold;'>标签名已存在，请输入一个唯一的标签名。</span>")
            return
        if ok and newText:
            label.setText(newText)
            label.setObjectName(newText)
            self.formRows[self.selectedGroup][self.selectedRowIndex].key = newText
            self.formRows[self.selectedGroup][self.selectedRowIndex].label = label

    def createFormRow(self, group, key, value):
        form = self.groupLayouts.get(group)
        if not form:
            return FormRow()
        rowWidget = QWidget()
        rowLayout = QFormLayout(rowWidget)
        label = PressableLabel(key, self)
        label.setObjectName(key)
        edit = QLineEdit(value, self)
        rowLayout.addRow(label, edit)
        self.bind_label_events(label)
        entry = IniEntry(group, key, value)
        self.bind_edit_events(edit, entry)
        return FormRow(rowWidget, label, edit, key)

    def parse_init_file(self, filepath):
        self.orderedEntries.clear()
        file = QFile(filepath)
        if not file.open(QIODevice.ReadOnly | QIODevice.Text):
            QMessageBox.warning(self, "错误", "<span style='color: green; font-weight: bold;'>无法打开 ini 文件</span>")
            return
        in_stream = QTextStream(file)
        currentGroup = ""

        while not in_stream.atEnd():
            line = in_stream.readLine()  # Returns a QString
            line = line.strip()  # Convert to str and remove whitespace
            if line.startswith(";") or not line:
                continue
            if line.startswith("[") and line.endswith("]"):
                currentGroup = line[1:-1].strip()  # Extract group name and strip
            else:
                eqIndex = line.find('=')  # Use find instead of indexOf
                if eqIndex != -1:
                    key = line[:eqIndex].strip()
                    value = line[eqIndex + 1:].strip()
                    value = value.replace("[", "").replace("]", "")  # Remove brackets
                    self.orderedEntries.append(IniEntry(currentGroup, key, value))
        file.close()

    def update_params(self):
        # 收集所有 QLineEdit 的当前值并更新 ComparisonWorker
        for group in self.groupOrder:
            if group not in self.formRows:
                continue
            for row in self.formRows[group]:
                value = row.edit.text()
                entry = IniEntry(group, row.key, value)
                self._apply_entry(entry)
        config_manager.batch_changed.emit()

    def save_settings(self):
        self.update_params()
        file = QFile(self.filename)
        if not file.open(QIODevice.WriteOnly | QIODevice.Text):
            QMessageBox.warning(self, "错误", "<span style='color: green; font-weight: bold;'>无法写入 ini 文件</span>")
            return
        out_stream = QTextStream(file)
        lastGroup = ""
        for group in self.groupOrder:
            if group not in self.formRows:
                continue
            if group != lastGroup:
                if lastGroup:
                    out_stream << "\n"
                out_stream << "[" << group << "]\n"
                lastGroup = group
            for row in self.formRows[group]:
                value = row.edit.text()
                out_stream << row.key << " = [" << value << "]\n"
        file.close()
        QMessageBox.information(self, "成功", "<span style='color: green; font-weight: bold;'>设置已保存</span>")

    def find_row_index(self, group, key):
        if group not in self.formRows:
            return -1
        rows = self.formRows[group]
        for i, row in enumerate(rows):
            if row.key == key:
                return i
        return -1

    def hightlight_selected_row(self):
        if self.selectedGroup not in self.formRows:
            return
        self.selectedRowIndex = min(self.selectedRowIndex, len(self.formRows[self.selectedGroup]) - 1) if self.selectedRowIndex >= 0 else -1
        if self.selectedRowIndex < 0:
            return
        self.clear_old_group_style()
        form = self.groupLayouts.get(self.selectedGroup)
        if not form:
            return
        labelItem = form.itemAt(self.selectedRowIndex, QFormLayout.LabelRole)
        if labelItem:
            labelWidget = labelItem.widget()
            if labelWidget:
                labelWidget.setStyleSheet(Settings.LABEL_SELECTED_STYLE)  # 示例高亮样式

    def clear_old_group_style(self):
        form = self.groupLayouts.get(self.selectedGroup)
        if not form:
            return
        for i in range(form.rowCount()):
            labelItem = form.itemAt(i, QFormLayout.LabelRole)
            if labelItem:
                labelWidget = labelItem.widget()
                if labelWidget:
                    labelWidget.setStyleSheet("")

    def refresh_group_layout(self, group):
        form = self.groupLayouts.get(group)
        if not form:
            return
        for row in self.formRows.get(group, []):
            if row.label:
                row.label.setParent(None)
            if row.edit:
                row.edit.setParent(None)
        for i in range(form.rowCount() - 1, -1, -1):
            form.removeRow(i)
        for row in self.formRows.get(group, []):
            form.addRow(row.label, row.edit)
        for row in self.formRows.get(group, []):
            if row.label:
                row.label.setParent(form.parentWidget())
            if row.edit:
                row.edit.setParent(form.parentWidget())
        self.selectedRowIndex = min(self.selectedRowIndex, form.rowCount() - 1)
        if 0 <= self.selectedRowIndex < form.rowCount():
            self.hightlight_selected_row()
        else:
            self.selectedRowIndex = -1

    def add_key_value_row(self):
        for i in range(1, 1000):  # 防止无限循环
            newKey = f"new_key_{i}"
            if not any(row.key == newKey for row in self.formRows.get(self.selectedGroup, [])):
                row = self.createFormRow(self.selectedGroup, newKey, "value")
                self.formRows.setdefault(self.selectedGroup, []).append(row)
                self.refresh_group_layout(self.selectedGroup)
                break

    def insert_key_value_row(self):
        if self.selectedRowIndex < 0:
            return
        for i in range(1, 1000):
            newKey = f"new_key_{i}"
            if not any(row.key == newKey for row in self.formRows.get(self.selectedGroup, [])):
                row = self.createFormRow(self.selectedGroup, newKey, "value")
                if self.selectedRowIndex >= len(self.formRows[self.selectedGroup]):
                    self.formRows[self.selectedGroup].append(row)
                else:
                    self.formRows[self.selectedGroup].insert(self.selectedRowIndex, row)
                self.refresh_group_layout(self.selectedGroup)
                break

    def delete_key_value_row(self):
        if self.selectedRowIndex < 0:
            return
        self.selectedRowIndex = min(self.selectedRowIndex, len(self.formRows.get(self.selectedGroup, [])) - 1)
        if self.selectedGroup in self.formRows and 0 <= self.selectedRowIndex < len(self.formRows[self.selectedGroup]):
            del self.formRows[self.selectedGroup][self.selectedRowIndex]
        self.refresh_group_layout(self.selectedGroup)

    def move_key_value_row(self, direction):
        if self.selectedRowIndex < 0 or self.selectedRowIndex >= len(self.formRows.get(self.selectedGroup, [])):
            return
        if (direction == -1 and self.selectedRowIndex == 0) or (direction == 1 and self.selectedRowIndex == len(self.formRows[self.selectedGroup]) - 1):
            return
        targetIndex = self.selectedRowIndex + direction
        if self.selectedGroup in self.formRows and 0 <= targetIndex < len(self.formRows[self.selectedGroup]):
            self.formRows[self.selectedGroup][self.selectedRowIndex], self.formRows[self.selectedGroup][targetIndex] = (
                self.formRows[self.selectedGroup][targetIndex], self.formRows[self.selectedGroup][self.selectedRowIndex]
            )
            self.selectedRowIndex = targetIndex
        self.refresh_group_layout(self.selectedGroup)

    def on_close(self):
        self.close()
        print("Page parameters closed")

if __name__ == "__main__":
    app = QApplication(sys.argv)
    window = PageParameters()
    window.show()
    sys.exit(app.exec()) 

File: cpp/src/page_parameters.cpp
#include "pva/page_parameters.hpp"
#include "pva/app_signals.hpp"
#include "pva/config_manager.hpp"
#include "ui_PageParameters.h"
#include <algorithm>
#include <utility>
#include <QEvent>
#include <QFile>
#include <QFormLayout>
#include <QLabel>
#include <QLineEdit>
#include <QMessageBox>
#include <QInputDialog>
#include <QSaveFile>
#include <QScrollArea>
#include <QTabBar>
#include <QTextStream>
#include <QVBoxLayout>

namespace pva
{
    namespace
    {
        QString displayValue(QString value)
        {
            value = value.trimmed();
            if (value.startsWith('[') && value.endsWith(']'))
                value = value.mid(1, value.size() - 2).trimmed();
            return value;
        }

        QString storedValue(QString value)
        {
            // 参数页只编辑纯值；写盘时恢复项目约定的 [value] 格式。
            return '[' + displayValue(std::move(value)) + ']';
        }

        QFormLayout *formForPage(QWidget *page)
        {
            if (!page)
                return nullptr;
            auto *container = page->findChild<QWidget *>("formContainer");
            return container ? qobject_cast<QFormLayout *>(container->layout()) : nullptr;
        }
    }

    PageParameters::PageParameters(QString path, QWidget *parent)
        : BasePage(parent), ui_(std::make_unique<Ui::PageParameters>()), configPath_(std::move(path))
    {
        initializePage([this] { ui_->setupUi(this); });
    }

    void PageParameters::initializeState() {}

    void PageParameters::setupPageUi()
    {
        ui_->tabWidget->setStyleSheet(
            "QTabWidget::pane { border: none; background-color: transparent; }");
    }

    void PageParameters::bindEvents()
    {
        connect(ui_->btn_load_parm, &QPushButton::clicked, this, &PageParameters::load);
        connect(ui_->btn_cancel_parm, &QPushButton::clicked, this, [this]
                { loadFromDisk(); });
        connect(ui_->btn_save_parm, &QPushButton::clicked, this, &PageParameters::save);
        connect(ui_->btn_add, &QPushButton::clicked, this, &PageParameters::addRow);
        connect(ui_->btn_insert, &QPushButton::clicked, this, &PageParameters::insertRow);
        connect(ui_->btn_delete, &QPushButton::clicked, this, &PageParameters::deleteRow);
        connect(ui_->btn_up, &QPushButton::clicked, this, &PageParameters::moveRowUp);
        connect(ui_->btn_down, &QPushButton::clicked, this, &PageParameters::moveRowDown);
        connect(ui_->tabWidget, &QTabWidget::currentChanged, this, [this](int)
                {
                    selectedGroup_ = ui_->tabWidget->currentIndex() >= 0
                                         ? ui_->tabWidget->tabText(ui_->tabWidget->currentIndex())
                                         : QString{};
                    selectedRowIndex_ = -1;
                    selectRow(-1);
                    // Python Settings.WIDGET_TAB_STYLE：当前参数组使用粉色左边线和蓝灰背景。
                    ui_->tabWidget->tabBar()->setStyleSheet(
                        "QTabBar::tab:selected {"
                        "border-left: 2px solid rgb(255, 121, 198);"
                        "background-color: rgb(86, 99, 136);"
                        "color: rgb(255, 255, 255);"
                        "}");
                });
    }

    void PageParameters::bindSignals()
    {
        connect(&AppSignals::instance(), &AppSignals::appClose, this, &QWidget::close);
    }

    void PageParameters::onReady()
    {
        loadFromDisk();
        emit ConfigManager::instance().batchChanged();
        ui_->tabWidget->tabBar()->setStyleSheet(
            "QTabBar::tab:selected {"
            "border-left: 2px solid rgb(255, 121, 198);"
            "background-color: rgb(86, 99, 136);"
            "color: rgb(255, 255, 255);"
            "}");
    }
    PageParameters::~PageParameters() = default;
    void PageParameters::clearTabs()
    {
        selectedRowIndex_ = -1;
        selectedGroup_.clear();
        groupLayouts_.clear();
        formRows_.clear();
        groupOrder_.clear();
        while (ui_->tabWidget->count())
            delete ui_->tabWidget->widget(0);
    }
    void PageParameters::load()
    {
        if (loadFromDisk())
        {
            ConfigManager::instance().load(configPath_);
            QMessageBox::information(this, "Load", "Settings loaded successfully.");
        }
    }

    bool PageParameters::loadFromDisk()
    {
        QFile file(configPath_);
        if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
        {
            QMessageBox::warning(this, "Load", file.errorString());
            return false;
        }
        clearTabs();
        QWidget *page = nullptr;
        QFormLayout *form = nullptr;
        QTextStream in(&file);
        while (!in.atEnd())
        {
            QString line = in.readLine().trimmed();
            if (line.startsWith('[') && line.endsWith(']'))
            {
                page = new QWidget;
                page->setProperty("section", line.mid(1, line.size() - 2));

                // Keep the same hierarchy as the Python page.  In particular,
                // the scroll area's frame must not become an extra box below
                // the tabs.
                auto *outerLayout = new QVBoxLayout(page);
                auto *scroll = new QScrollArea;
                scroll->setWidgetResizable(true);
                scroll->setStyleSheet("QScrollArea { border: none; }");

                auto *content = new QWidget;
                content->setObjectName("formContainer");
                content->setStyleSheet(
                    "QWidget#formContainer {"
                    "border: 2px solid rgb(70, 80, 110);"
                    "background: transparent;"
                    "padding: 0px;"
                    "}");
                form = new QFormLayout(content);
                scroll->setWidget(content);
                outerLayout->addWidget(scroll);
                const QString group = page->property("section").toString();
                ui_->tabWidget->addTab(page, group);
                groupLayouts_.insert(group, form);
                groupOrder_.append(group);
            }
            else if (form && !line.isEmpty() && !line.startsWith('#') &&
                     !line.startsWith(';') && line.contains('='))
            {
                auto parts = line.split('=');
                QString key = parts.takeFirst().trimmed(), value = displayValue(parts.join('='));
                const QString group = page->property("section").toString();
                formRows_[group].append(createRow(group, key, value));
            }
        }
        for (const QString &group : std::as_const(groupOrder_))
            refreshGroupLayout(group);
        if (ui_->tabWidget->count() > 0)
        {
            selectedGroup_ = ui_->tabWidget->tabText(ui_->tabWidget->currentIndex());
            selectedRowIndex_ = -1;
        }
        return true;
    }
    void PageParameters::save()
    {
        QSaveFile file(configPath_);
        if (!file.open(QIODevice::WriteOnly | QIODevice::Text))
        {
            QMessageBox::warning(this, "Save", file.errorString());
            return;
        }
        QTextStream out(&file);
        for (const QString &group : std::as_const(groupOrder_))
        {
            out << '[' << group << "]\n";
            for (const FormRow &row : formRows_.value(group))
                if (row.edit)
                    out << row.key << " = " << storedValue(row.edit->text()) << "\n";
            out << "\n";
        }
        if (!file.commit())
        {
            QMessageBox::warning(this, "Save", file.errorString());
            return;
        }
        ConfigManager::instance().load(configPath_);
        emit configurationSaved();
        QMessageBox::information(this, "Save", "Settings saved successfully.");
    }

    QFormLayout *PageParameters::currentForm() const
    {
        return groupLayouts_.value(selectedGroup_, formForPage(ui_->tabWidget->currentWidget()));
    }

    QLineEdit *PageParameters::selectedEditor() const
    {
        const auto rows = formRows_.value(selectedGroup_);
        if (selectedRowIndex_ < 0 || selectedRowIndex_ >= rows.size())
            return nullptr;
        return rows.at(selectedRowIndex_).edit;
    }

    QLineEdit *PageParameters::editorAt(QFormLayout *form, int row) const
    {
        auto *item = form && row >= 0 && row < form->rowCount()
                         ? form->itemAt(row, QFormLayout::FieldRole)
                         : nullptr;
        return item ? qobject_cast<QLineEdit *>(item->widget()) : nullptr;
    }

    void PageParameters::registerRow(QLabel *label, QLineEdit *edit)
    {
        label->setBuddy(edit);
        label->installEventFilter(this);
        edit->installEventFilter(this);
    }

    void PageParameters::selectRow(int row)
    {
        auto &rows = formRows_[selectedGroup_];
        selectedRowIndex_ = row >= 0 && row < rows.size() ? row : -1;
        for (int index = 0; index < rows.size(); ++index)
            if (rows[index].label)
                rows[index].label->setStyleSheet(index == selectedRowIndex_
                    ? "border-left: 2px solid rgb(189, 147, 249);"
                      "background-color: rgb(86, 99, 136);"
                      "color: rgb(255, 255, 255);"
                    : QString{});
    }

    int PageParameters::rowIndexFor(const QObject *widget) const
    {
        const auto rows = formRows_.value(selectedGroup_);
        for (int index = 0; index < rows.size(); ++index)
            if (rows[index].label == widget || rows[index].edit == widget)
                return index;
        return -1;
    }

    bool PageParameters::eventFilter(QObject *watched, QEvent *event)
    {
        auto *edit = qobject_cast<QLineEdit *>(watched);
        auto *label = qobject_cast<QLabel *>(watched);
        if (!edit && label)
            edit = qobject_cast<QLineEdit *>(label->buddy());
        if (edit && (event->type() == QEvent::FocusIn || event->type() == QEvent::MouseButtonPress ||
                     event->type() == QEvent::MouseButtonDblClick))
            selectRow(rowIndexFor(edit));
        if (edit && label && event->type() == QEvent::MouseButtonDblClick)
        {
            bool accepted = false;
            const QString oldKey = edit->property("key").toString();
            const QString newKey = QInputDialog::getText(this, "Parameter", "Key:",
                                                         QLineEdit::Normal, oldKey, &accepted).trimmed();
            auto *form = currentForm();
            bool duplicate = false;
            for (int row = 0; form && row < form->rowCount(); ++row)
                if (auto *other = editorAt(form, row); other && other != edit &&
                                                       other->property("key").toString() == newKey)
                    duplicate = true;
            if (accepted && !newKey.isEmpty() && !duplicate)
            {
                if (selectedRowIndex_ >= 0 && selectedRowIndex_ < formRows_[selectedGroup_].size())
                    formRows_[selectedGroup_][selectedRowIndex_].key = newKey;
                edit->setProperty("key", newKey);
                label->setText(newKey);
            }
            else if (accepted && duplicate)
                QMessageBox::warning(this, "Parameter", "The parameter key already exists.");
            return true;
        }
        return QWidget::eventFilter(watched, event);
    }

    PageParameters::FormRow PageParameters::createRow(const QString &, const QString &key, const QString &value)
    {
        auto *edit = new QLineEdit(value, this);
        edit->setProperty("key", key);
        auto *label = new QLabel(key, this);
        label->setObjectName(key);
        registerRow(label, edit);
        return FormRow{key, label, edit};
    }

    QString PageParameters::nextKey(const QString &group) const
    {
        for (int index = 1; index < 1000; ++index)
        {
            const QString candidate = QString("new_key_%1").arg(index);
            const auto rows = formRows_.value(group);
            if (std::none_of(rows.cbegin(), rows.cend(), [&candidate](const FormRow &row)
                             { return row.key == candidate; }))
                return candidate;
        }
        return {};
    }

    void PageParameters::refreshGroupLayout(const QString &group)
    {
        auto *form = groupLayouts_.value(group);
        if (!form)
            return;
        while (form->rowCount() > 0)
        {
            auto taken = form->takeRow(0);
            delete taken.labelItem;
            delete taken.fieldItem;
        }
        for (const FormRow &row : std::as_const(formRows_[group]))
            form->addRow(row.label, row.edit);
        if (group == selectedGroup_)
            selectRow(selectedRowIndex_);
    }

    void PageParameters::addRow()
    {
        const QString key = nextKey(selectedGroup_);
        if (key.isEmpty())
            return;
        formRows_[selectedGroup_].append(createRow(selectedGroup_, key, "value"));
        refreshGroupLayout(selectedGroup_);
    }

    void PageParameters::insertRow()
    {
        if (selectedRowIndex_ < 0)
            return;
        const QString key = nextKey(selectedGroup_);
        if (key.isEmpty())
            return;
        formRows_[selectedGroup_].insert(selectedRowIndex_, createRow(selectedGroup_, key, "value"));
        refreshGroupLayout(selectedGroup_);
    }

    void PageParameters::deleteRow()
    {
        auto &rows = formRows_[selectedGroup_];
        if (selectedRowIndex_ < 0 || selectedRowIndex_ >= rows.size())
            return;
        const FormRow removed = rows.takeAt(selectedRowIndex_);
        selectedRowIndex_ = std::min(selectedRowIndex_, static_cast<int>(rows.size()) - 1);
        refreshGroupLayout(selectedGroup_);
        delete removed.label;
        delete removed.edit;
    }

    void PageParameters::moveSelectedRow(int delta)
    {
        auto &rows = formRows_[selectedGroup_];
        const int target = selectedRowIndex_ + delta;
        if (selectedRowIndex_ < 0 || target < 0 || target >= rows.size())
            return;
        rows.swapItemsAt(selectedRowIndex_, target);
        selectedRowIndex_ = target;
        refreshGroupLayout(selectedGroup_);
    }

    void PageParameters::moveRowUp() { moveSelectedRow(-1); }
    void PageParameters::moveRowDown() { moveSelectedRow(1); }
}
